crates/flipperzero/src/toolbox/sha256.rs

```diff
@@ -0,0 +1,118 @@
+use core::fmt;
+
+use digest::{
+    block_buffer::Eager,
+    core_api::{Buffer, BufferKindUser, CoreWrapper, FixedOutputCore, UpdateCore},
+    crypto_common::{AlgorithmName, Block, BlockSizeUser},
+    typenum::{Unsigned, U32, U64},
+    HashMarker, Output, OutputSizeUser, Reset,
+};
+use flipperzero_sys as sys;
+
+/// The [SHA-256 hash function][1].
+///
+/// Equivalent to [`sha2::Sha256`].
+///
+/// [1]: https://en.wikipedia.org/wiki/SHA-2
+pub type Sha256 = CoreWrapper<Sha256Core>;
+
+/// Core block-level SHA-256 hasher.
+pub struct Sha256Core {
+    state: sys::sha256_context,
+}
+
+impl HashMarker for Sha256Core {}
+
+impl BlockSizeUser for Sha256Core {
+    type BlockSize = U64;
+}
+
+impl BufferKindUser for Sha256Core {
+    type BufferKind = Eager;
+}
+
+impl OutputSizeUser for Sha256Core {
+    type OutputSize = U32;
+}
+
+impl Default for Sha256Core {
+    #[inline]
+    fn default() -> Self {
+        let mut state = sys::sha256_context {
+            total: [0; 2],
+            state: [0; 8],
+            wbuf: [0; 16],
+        };
+        unsafe { sys::sha256_start(&mut state) };
+        Self { state }
+    }
+}
+
+impl UpdateCore for Sha256Core {
+    #[inline]
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        for block in blocks {
+            self.state.total[0] += Self::BlockSize::U32;
+            if self.state.total[0] < Self::BlockSize::U32 {
+                self.state.total[1] += 1;
+            }
+
+            unsafe {
+                sys::memcpy(
+                    self.state.wbuf.as_mut_ptr().cast(),
+                    block.as_ptr().cast(),
+                    Self::BlockSize::U32,
```

Comment:
I'm assuming this evaluates to `64u32` Would `Self::BlockSize::USIZE` match the signature better here?

Reply:
If we used [`core::ptr::copy_nonoverlapping`](https://doc.rust-lang.org/core/ptr/fn.copy_nonoverlapping.html) instead then yes. But `sys::memcpy` takes [`core::ffi::c_uint`](https://doc.rust-lang.org/stable/core/ffi/type.c_uint.html) as its third argument, which is a `u32`.

Reply:
Makes sense. I have to remember to check the Flipper APIs vs. the Linux ones.